sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TableChange.java

```diff
@@ -787,4 +806,81 @@ public int hashCode() {
       return Arrays.hashCode(clusteringColumns);
     }
   }
+
+  /** A TableChange to alter table and add a constraint. */
+  final class AddConstraint implements TableChange {
+    private final Constraint constraint;
+    private final boolean validate;
```

Comment:
During the discussion, the SPIP design evolved to be like this:

```
class AddConstraint implements TableChange {
  private final Constraint constraint;
  private final String validatedTableVersion; // if known and was validated 
  …
}
```

If Spark validated the constraint, it should set the validation status in the constraint to `VALID`. The table version will be there, if known.

```
interface Table {
  …

  // reports the current table version, if supported
  default String currentVersion() { return null; }

  …
}
```

Reply:
TBH I am a bit confused on this one. Shall we make it simple and let the external table catalog to keep trace the `validatedTableVersion`?

Reply:
The problem is that will construct a scan to do the validation. We need a way to pass which version of the table was validated back to the connector so that the connector can check changes that happened concurrently.